### PR TITLE
jv-fruit-shop-tests-solution

### PR DESCRIPTION
**Pull request by BolsunovaNataliia**

### CODE REVIEW

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Test
+    public void addElementsToStorage() {
+        List<String> lines = new ArrayList<String>();
+        lines.add("type,fruit,quantity");
+        lines.add("b,banana,20");
+        lines.add("b,apple,90");
+        lines.add("s,banana,100");
+        lines.add("p,banana,40");
+        lines.add("s,apple,10");
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:38]: should be initialised in the BeforeAll method and create a constant

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        FruitMapper fruitMapper = new FruitMapperImpl();
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:39]: create a constant

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        );
+        transactionProcessor.process(fruitTransactions);
+        Map<String, Integer> actual = stock.getData();
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void getData_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:54]: create a constant

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                "banana", 80,
+                "apple", 120
+        );
+        assertEquals(expected, stock.getData());
+    }
+
+    @Test
+    public void getData_containsKey_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:65]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                "banana", 80,
+                "apple", 120
+        );
+        assertEquals(expected.containsKey("banana"),
+                stock.getData().containsKey("banana"));
+    }
+
+    @Test
+    public void getData_containsKey_notOk() {
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:76]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        assertFalse(stock.getData().containsKey("banana"));
+    }
+
+    @Test
+    public void getData_getValue_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:83]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        lines.add("type,fruit,quantity");
+        lines.add("b,banana,20");
+        lines.add("b,apple,90");
+        lines.add("s,banana,100");
+        lines.add("p,banana,40");
+        lines.add("s,apple,10");
+        FruitMapper fruitMapper = new FruitMapperImpl();
+        List<FruitTransaction> fruitTransactions = fruitMapper.mapData(lines);
+        TransactionProcessor transactionProcessor = new TransactionProcessorImpl(stock);
+        Map<String, Integer> expected = Map.of(
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:42]: create a constant and initialise it in the method

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                "apple", 120
+        );
+        assertEquals(expected, stock.getData());
+    }
+
+    @Test
+    public void getData_containsKey_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
+        Map<String, Integer> expected = Map.of(
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:66]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                "banana", 80,
+                "apple", 120
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:68]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        assertFalse(stock.getData().containsKey("banana"));
+    }
+
+    @Test
+    public void getData_getValue_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
+        Map<String, Integer> expected = Map.of(
+                "banana", 80,
+                "apple", 120
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:86]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Test
+    public void getData_containsValue_notOk() {
+        stock.putData("apple", 120);
+        assertFalse(stock.getData().containsValue(80));
+    }
+
+    @Test
+    public void getData_getValue_notOk() {
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:99]: the same

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        Map<String, Integer> expected = Map.of(
+                "banana", 80,
+                "apple", 120
+        );
+        assertEquals(expected.get("banana"), stock.getData().get("banana"));
+    }
+
+    @Test
+    public void getData_containsValue_notOk() {
+        stock.putData("apple", 120);
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/data/StockTest.java:93]: the same

File: src/test/java/core/basesyntax/io/CsvFileReaderTest.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class CsvFileReaderTest {
+    private static final String VALID_PATH = "src/test/resources/file.csv";
+    private static final String INVALID_PATH = "src20/test/resources/file.csv";
+    private static final String FILE_EMPTY_PATH = "src/test/resources/file_empty.csv";
+    private static final String FILE_HAS_ONLY_HEAD_PATH =
+            "src/test/resources/file_has_only_head.csv";
+    private ReaderFromFile readerFromFile = new CsvFileReader();
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/io/CsvFileReaderTest.java:15]: should be initialised in the BeforeAll method

File: src/test/java/core/basesyntax/io/CsvFileReaderTest.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package core.basesyntax.io;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertIterableEquals;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class CsvFileReaderTest {
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/io/CsvFileReaderTest.java:9]: another null test is required

File: src/test/java/core/basesyntax/io/CsvFileWriterTest.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package core.basesyntax.io;
+
+import static org.junit.Assert.assertThrows;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class CsvFileWriterTest {
+    private static final String INVALID_PATH = "src20/test/resources/report.csv";
+    private WriterToFile writer = new CsvFileWriter();
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/io/CsvFileWriterTest.java:10]: should be initialised in the method

File: src/test/java/core/basesyntax/io/CsvFileWriterTest.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
+package core.basesyntax.io;
+
+import static org.junit.Assert.assertThrows;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class CsvFileWriterTest {
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/io/CsvFileWriterTest.java:8]: more tests for Null/valid test/empty file are needed

File: src/test/java/core/basesyntax/service/FruitMapperImplTest.java [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    private static final int QUANTITY_BANANA_BALANCE = 20;
+    private static final int ZERO_INDEX = 0;
+    private static final String CODE_OPERATION_BALANCE = "b";
+    private static final String HEAD_INPUT = "type,fruit,quantity";
+    private static final String FIRST_INVALID_ROW = "b,null,20";
+    private static final String FIRST_VALID_ROW = "b,banana,20";
+    private static final String FRUIT_BANANA = "banana";
+    private static final String SECOND_VALID_ROW = "b,apple,100";
+    private FruitMapper mapper = new FruitMapperImpl();
+    private List<String> lines = new ArrayList<>();
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/service/FruitMapperImplTest.java:21]: should be initialised in the method

File: src/test/java/core/basesyntax/service/ReportServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+public class ReportServiceTest {
+    private static final String HEAD_REPORT = "fruit,quantity";
+    private static final String REGEX_COMMA = ",";
+    private static final String ROW_OF_INPUT_DATA = "apple,110";
+    private static final int ROW_OF_INPUT_DATA_INDEX = 1;
+    private static final int HEAD_REPORT_INDEX = 0;
+    private static final Map<String, Integer> data = Map.of(
+            "apple", 110,
+            "banana", 90
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/service/ReportServiceTest.java:17]: should be initialised in the method

File: src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    private static final String FRUIT_BANANA = "banana";
+    private BalanceDataHandler balanceDataHandler = new BalanceDataHandler();
+    private FruitTransaction validTransaction =
+            new FruitTransaction(FruitTransaction.Operation.BALANCE,
+                    FRUIT_BANANA, QUANTITY_BANANA);
+    private FruitTransaction invalidTransaction =
+            new FruitTransaction(FruitTransaction.Operation.PURCHASE,
+                    FRUIT_BANANA, QUANTITY_BANANA);
+    private Map<String, Integer> validData = Map.of(
+            FRUIT_BANANA,QUANTITY_BANANA
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java:19]: should be initialised in the method

File: src/test/java/core/basesyntax/strategy/ReturnDataHandlerTest.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    private static final String FRUIT_APPLE = "apple";
+    private ReturnDataHandler returnDataHandler = new ReturnDataHandler();
+    private FruitTransaction validTransaction =
+            new FruitTransaction(FruitTransaction.Operation.RETURN, FRUIT_APPLE, QUANTITY_APPLE);
+    private FruitTransaction invalidTransaction =
+            new FruitTransaction(FruitTransaction.Operation.PURCHASE, FRUIT_APPLE, QUANTITY_APPLE);
+    private Map<String, Integer> validData = Map.of(
+            FRUIT_APPLE,QUANTITY_APPLE
+    );
+
```
Comment (OlexVoit) [inline src/test/java/core/basesyntax/strategy/ReturnDataHandlerTest.java:19]: should be initialised in the method

**Review by OlexVoit — CHANGES_REQUESTED** [review]
Good job!
The main problem is that objects are not initialized in the methods. 
There is also code duplication.
Doesn't have tests for two classes in the Strategy package and any tests.

File: src/test/java/core/basesyntax/data/StockTest.java (hunk 1 above)
Comment (BolsunovaNataliia) [reply src/test/java/core/basesyntax/data/StockTest.java:38]: I can't save it as a constant because I want to be able to change it.

File: src/test/java/core/basesyntax/data/StockTest.java (hunk 3 above)
Comment (BolsunovaNataliia) [reply src/test/java/core/basesyntax/data/StockTest.java:54]: I can't save it as a constant because I want to be able to change it.

File: src/test/java/core/basesyntax/data/StockTest.java (hunk 4 above)
Comment (BolsunovaNataliia) [reply src/test/java/core/basesyntax/data/StockTest.java:65]: I can't save it as a constant because I want to be able to change it.

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.junit.Test;
+import org.junit.jupiter.api.BeforeEach;
+
+public class StockTest {
+    private static final String HEAD_INPUT = "type,fruit,quantity";
+    private Stock stock = new Stock();
+    private List<String> lines;
```
Comment (d1sam) [inline src/test/java/core/basesyntax/data/StockTest.java:20]:
```suggestion
    private List<String> inputData;
```

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    @Test
+    public void getData_containsKey_Ok() {
+        stock.putData("banana", 80);
+        stock.putData("apple", 120);
+        Map<String, Integer> expected = Map.of(
+                "banana", 80,
+                "apple", 120
+        );
+        assertEquals(expected.containsKey("banana"),
+                stock.getData().containsKey("banana"));
```
Comment (d1sam) [inline src/test/java/core/basesyntax/data/StockTest.java:68]:
```suggestion
        assertEquals(expected, stock.getData());
```

Comment (d1sam) [reply src/test/java/core/basesyntax/data/StockTest.java:68]: you can compare 2 maps, this test will be better, than contains.

File: src/test/java/core/basesyntax/data/StockTest.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Test
+    public void getData_containsKey_notOk() {
+        stock.putData("apple", 120);
+        assertFalse(stock.getData().containsKey("banana"));
```
Comment (d1sam) [inline src/test/java/core/basesyntax/data/StockTest.java:74]: do you test `containsKey` realization of java.util.Map?

File: src/test/java/core/basesyntax/io/CsvFileWriterTest.java [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+        report = List.of(
+                HEAD_REPORT,
+                FIRST_LINE_REPORT,
+                SECOND_LINE_REPORT
+        );
+        Assertions.assertTrue(report.contains(FIRST_LINE_REPORT));
+    }
+
+    @Test
+    void writeFile() {
```
Comment (d1sam) [inline src/test/java/core/basesyntax/io/CsvFileWriterTest.java:72]: make test name with expected state.

File: src/test/java/core/basesyntax/io/CsvFileWriterTest.java [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Test
+    void checkFirstLineReportExist_Ok() {
+        report = List.of(
+                HEAD_REPORT,
+                FIRST_LINE_REPORT,
+                SECOND_LINE_REPORT
+        );
+        Assertions.assertTrue(report.contains(FIRST_LINE_REPORT));
```
Comment (d1sam) [inline src/test/java/core/basesyntax/io/CsvFileWriterTest.java:58]:
```suggestion
       assertTrue(report.contains(FIRST_LINE_REPORT));
```

Comment (d1sam) [reply src/test/java/core/basesyntax/io/CsvFileWriterTest.java:58]: because you use `assertThrows` with static import

File: src/test/java/core/basesyntax/io/CsvFileWriterTest.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String value = reader.readLine();
+            while (value != null) {
+                actualReport.add(value);
+                value = reader.readLine();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(EXCEPTION_FIND_FILE_MESSAGE + VALID_PATH, e);
+        }
+        Assertions.assertEquals(report, actualReport);
```
Comment (d1sam) [inline src/test/java/core/basesyntax/io/CsvFileWriterTest.java:90]:
```suggestion
       assertEquals(report, actualReport);
```

Comment (d1sam) [reply src/test/java/core/basesyntax/io/CsvFileWriterTest.java:90]: because you use `assertThrows` with static import

File: src/test/java/core/basesyntax/service/FruitMapperImplTest.java [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+    private static final int QUANTITY_BANANA_BALANCE = 20;
+    private static final int ZERO_INDEX = 0;
+    private static final String CODE_OPERATION_BALANCE = "b";
+    private static final String HEAD_INPUT = "type,fruit,quantity";
+    private static final String FIRST_INVALID_ROW = "b,null,20";
+    private static final String FIRST_VALID_ROW = "b,banana,20";
+    private static final String FRUIT_BANANA = "banana";
+    private static final String SECOND_VALID_ROW = "b,apple,100";
+    private FruitMapper mapper;
+    private List<String> lines;
```
Comment (d1sam) [inline src/test/java/core/basesyntax/service/FruitMapperImplTest.java:21]:
```suggestion
    private List<String> inputData;
```

File: src/test/java/core/basesyntax/service/FruitMapperImplTest.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        lines.add(HEAD_INPUT);
+        Assertions.assertThrows(RuntimeException.class, () -> mapper.mapData(lines));
+    }
+
+    @Test
+    void listContainsHead_Ok() {
+        lines.add(HEAD_INPUT);
+        lines.add(FIRST_VALID_ROW);
+        Assertions.assertTrue(lines.contains(HEAD_INPUT));
+    }
```
Comment (d1sam) [inline src/test/java/core/basesyntax/service/FruitMapperImplTest.java:56]: do you test `contains` method of `List`?

File: src/test/java/core/basesyntax/service/ReportServiceTest.java [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+    void checkHeadReport_Ok() {
+        List<String> report = reportService.generateReport(data);
+        Assertions.assertEquals(HEAD_REPORT, report.get(HEAD_REPORT_INDEX));
+    }
+
+    @Test
+    void checkComma_Ok() {
+        List<String> report = reportService.generateReport(data);
+        Assertions.assertTrue(report.get(0).contains(REGEX_COMMA));
+    }
```
Comment (d1sam) [inline src/test/java/core/basesyntax/service/ReportServiceTest.java:48]: if `checkHeadReport_Ok` is ok, this test - ok, if `checkHeadReport_Ok` is failure this also be failure. That's why this test with `contains` is redundant.

Comment (d1sam) [inline src/test/java/core/basesyntax/service/ReportServiceTest.java:48]:
```suggestion
```

File: src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        Assertions.assertTrue(validTransaction.getOperation()
+                .equals(FruitTransaction.Operation.BALANCE));
+    }
+
+    @Test
+    void invalidBalanceOperationInTransaction_notOk() {
+        invalidTransaction =
+                new FruitTransaction(FruitTransaction.Operation.PURCHASE,
+                        FRUIT_BANANA, QUANTITY_BANANA);
+        Assertions.assertFalse(FruitTransaction.Operation.BALANCE
```
Comment (d1sam) [inline src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java:36]:
```suggestion
        Assertions.assertNotEquals(FruitTransaction.Operation.BALANCE, invalidTransaction.getOperation());
```

File: src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+    void setUp() {
+        balanceDataHandler = new BalanceDataHandler();
+    }
+
+    @Test
+    void validBalanceOperationInTransaction_Ok() {
+        validTransaction =
+                new FruitTransaction(FruitTransaction.Operation.BALANCE,
+                        FRUIT_BANANA, QUANTITY_BANANA);
+        Assertions.assertTrue(validTransaction.getOperation()
```
Comment (d1sam) [inline src/test/java/core/basesyntax/strategy/BalanceDataHandlerTest.java:27]: let's better use assert equals for better readability.

File: src/test/java/core/basesyntax/strategy/SupplyDataHandlerTest.java [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    void validProcessWithData_Ok() {
+        validData = Map.of(
+                FRUIT_BANANA,QUANTITY_BANANA);
+        validTransaction =
+                new FruitTransaction(FruitTransaction.Operation.BALANCE,
+                        FRUIT_BANANA, QUANTITY_BANANA);
+        Assertions.assertTrue(
+                validData.containsKey(validTransaction.getFruit()));
```
Comment (d1sam) [inline src/test/java/core/basesyntax/strategy/SupplyDataHandlerTest.java:48]: let's better use for this test `assertEquals` on `FRUIT_BANANA` and `validTransaction.getFruit()`. Because now you also test `containsKey` method.

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job , but some changes should be done. Also please provide a meaningful error messages in your tests. Also please make constants everywhere you use String literals.

**Review by bhdnchui — APPROVED** [review]
Good job!
